ui/raidboss/data/07-dt/ultimate/futures_rewritten.ts

```diff
@@ -1808,7 +1891,668 @@ const triggerSet: TriggerSet<Data> = {
     // ************************
     // P4 -- Duo
     // ************************
+    {
+      id: 'FRU P4 Akh Rhai',
+      // positions snapshot when hidden combatants are added, but they have unreliable names
+      // we can safely base this on the timing of the vfx effect that precedes the snapshot
+      type: 'GainsEffect',
+      netRegex: { effectId: '8E1', capture: false },
+      condition: (data) => data.phase === 'p4-dld',
+      delaySeconds: 4.7,
+      suppressSeconds: 1,
+      response: Responses.moveAway('alert'),
+    },
+    {
+      id: 'FRU P4 Akh Morn',
+      type: 'StartsUsing',
+      netRegex: { id: '9D6E', source: 'Oracle of Darkness', capture: false },
+      alertText: (_data, _matches, output) => output.stacks!(),
+      outputStrings: {
+        stacks: Outputs.stacks,
+      },
+    },
+    {
+      id: 'FRU PP4 Morn Afah',
+      type: 'StartsUsing',
+      netRegex: { id: '9D70', source: 'Oracle of Darkness', capture: false },
+      alertText: (_data, _matches, output) => output.stack!(),
+      outputStrings: {
+        stack: Outputs.getTogether,
+      },
+    },
+    // ***** Darklit Dragonsong *****
+    {
+      id: 'FRU P4 Darklit Dragonsong',
+      type: 'StartsUsing',
+      netRegex: { id: '9D6D', source: 'Oracle of Darkness', capture: false },
+      response: Responses.bigAoe(),
+    },
+    {
+      id: 'FRU P4 Darklit Stacks Collect',
+      type: 'GainsEffect',
+      netRegex: { effectId: '99D' }, // Spell-in-Waiting: Dark Water III
+      condition: (data) => data.phase === 'p4-dld',
+      run: (data, matches) => data.p4DarklitStacks.push(matches.target),
+    },
+    {
+      id: 'FRU P4 Darklit Tether + Cleave Collect',
+      type: 'Tether',
+      netRegex: { id: '006E' }, // Refulgent Chain
+      condition: (data) => data.phase === 'p4-dld',
+      run: (data, matches) => {
+        data.p4DarklitTetherCount++;
+        (data.p4DarklitTethers[matches.source] ??= []).push(matches.target);
+        (data.p4DarklitTethers[matches.target] ??= []).push(matches.source);
 
+        if (data.p4DarklitTetherCount === 4)
+          data.p4DarklitCleaves = data.party.partyNames.filter(
+            (name) => !(Object.keys(data.p4DarklitTethers).includes(name)),
+          );
+      },
+    },
+    // The logic for tether swaps, bait swaps, and possible stack swaps is fairly concise.
+    // It's not completely comprehensive (specifically, we can't determine which DPS need to flex
+    // and when for the cone baits), but otherwise it's accurate.  See inline comments.
+    {
+      id: 'FRU P4 Darklit Tower / Bait',
+      type: 'Tether',
+      netRegex: { id: '006E', capture: false }, // Refulgent Chain
+      condition: (data) => data.phase === 'p4-dld' && data.p4DarklitTetherCount === 4,
```

Comment:
Is there any possibility of a race with the collection trigger, given that both seem to be simultaneous?

Reply:
All 3 triggers fire on the same line, but they execute sequentially and synchronously (at least, that's what quisquous told me long ago when I was just getting started).  So `FRU P4 Darklit Tether + Cleave Collect` will run four times, incrementing `p4DarklitTetherCount` each time.  The `condition`s on `FRU P4 Darklit Tower / Bait` and `FRU P4 Darklit Cleave Stack` are only satisfied once the collect trigger has run four times, and they are executed sequentially after it.  Given that, I don't think a race condition could occur?

Reply:
Mm. I don't really like relying on an implicit "trigger order is also synchronous execution order", but I suppose you're right, that being the case this is safe as it is.